crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -363,6 +367,7 @@ unsafe impl<D: QueryData + 'static, F: QueryFilter + 'static> SystemParam for Qu
         world: UnsafeWorldCell<'w>,
         change_tick: Tick,
     ) -> Self::Item<'w, 's> {
+        let state = get_query_state(*state, world).unwrap();
         // SAFETY: We have registered all of the query's world accesses,
```

Comment:
I think this opens a soundness hole.  `Query` relies on having registered the correct access for the `QueryState`.  But if `QueryState` is a component, then user code could find the entity and replace the state with one that had different access (using `QueryBuilder` or `QueryState::transmute`).  

I'm not sure how to resolve this.  There is a similar issue with components-as-entities, although the consensus seemed to be that we'd still store the "real" component data elsewhere, which means we won't need to actually solve it.  But it feels like it might be an increasing problem as we everything-as-entities.  Maybe we need some kind of unsafe component that requires unsafe code to modify or insert?

Reply:
The simplest solution we've discussed prior would be rather than making `QueryState` itself a `Component`, instead we would have a non-`pub` wrapper `Component` that stores the `QueryState`. So user code can't accidentally break query soundness, but they still could *purposefully* if they can figure out the `ComponentId` of the wrapper component.

Reply:
> The simplest solution we've discussed prior would be rather than making `QueryState` itself a `Component`, instead we would have a non-`pub` wrapper `Component` that stores the `QueryState`. So user code can't accidentally break query soundness, but they still could _purposefully_ if they can figure out the `ComponentId` of the wrapper component.

Oh, good idea!  Yeah, even if they get the `ComponentId`, all they could do without `unsafe` is *remove* the component, which is still sound.  I must have been mixing this up with components-as-entities, where we want a `pub` type so that user code can *read* the data.

Reply:
How they can replace the state? EntityMut or similar stuff can not modify the structure of the entity.

Reply:
> How they can replace the state? EntityMut or similar stuff can not modify the structure of the entity. Edit: oh mem::replace stuff

Yup!  And also `EntityWorldMut::insert` and `EntityCommands::insert` will overwrite existing values, so even using immutable components here wouldn't help.  But using a non-`pub` type should do it!